design/csi-snapshots.md

```diff
@@ -0,0 +1,192 @@
+
+By supporting the CSI snapshot API, Velero can extend it's support to any CSI driver, without requiring a Velero-specific plugin be written, easing the development burden on providers while also reaching more end users.
+
+## High-Level Design
+
+In order to support CSI's snapshot API, Velero must interact with the [`VolumeSnapshot`][2] and [`VolumeSnapshotContent`][3] CRDs.
+These act as requests to the CSI driver to perform a snapshot on the underlying provider's volume.
+This can largely be accomplished with Velero `BackupItemAction` and `RestoreItemAction` plugins that operate on these CRDs.
+
+Additionally, changes to the Velero server and client code are necessary to track `VolumeSnapshots` that are associated with a given backup, similarly to how Velero tracks its own [`volume.Snapshot`][4] type.
+Tracking these is important for allowing users to see what is in their backup, and provides parity for the existing `volume.Snapshot` and [`PodVolumeBackup`][5] types.
+This is also done to retain the object store as Velero's source of truth, without having to query the Kubernetes API server for associated `VolumeSnapshot`s.
+
+`velero backup describe --details` will use the stored VolumeSnapshots to list CSI snapshots included in the backup to the user.
+
+## Detailed Design
+
+### Resource Plugins
+
+A set of [prototype][6] plugins was developed that informed this design.
+
+The plugins will be as follows:
+
+* A `BackupItemAction` for `PersistentVolumeClaim`s, named `velero.io/csi-pvc` 
+The associated PV will be queried and checked for the presence of `PersistentVolume.Spec.PersistentVolumeSource.CSI`. (See the "Snapshot Mechanism Selection" section below)
+If this field is `nil`, then the plugin will return early without taking action.
+Create a `VolumeSnapshot.snapshot.storage.k8s.io` object from the PVC.
+Label the `VolumeSnapshot` object with the [`velero.io/backup-name`][10] label for ease of lookup later.
+The CSI controllers will create a `VolumeSnapshotContent.snapshot.storage.k8s.io` object associated with the `VolumeSnapshot`.
+Associated `VolumeSnapshotContent` objects will be retrieved and updated with the [`velero.io/backup-name`][10] label for ease of lookup later.
+`velero.io/volume-snapshot-name` will be applied as a label to the PVC so that the `VolumeSnapshot` can be found easily for restore.
+`VolumeSnapshot`, `VolumeSnapshotContent`, and `VolumeSnapshotClass` objects would be returned as additional items to be backed up. GitHub issue [1566][18] represents this work.
+The `VolumeSnapshotContent.Spec.VolumeSnapshotSource.SnapshotHandle` field is the link to the underlying platform's snapshot, and must be preserved for restoration.
+
+The plugin will _not_ wait for the `VolumeSnapshot.Status.IsReady` field to be `true` before returning.
+This maintains current Velero behavior, though may not be desirable for all storage providers.
+
+* A `RestoreItemAction` for `VolumeSnapshotContent` objects, named `velero.io/csi-vsc`.
```

Comment:
I'm wondering if some of this code belongs in the backup sync controller - so prior to even executing a restore, we'd create the CSI CR's in the cluster to represent the snapshots. This would be more consistent with how we're handing PodVolumeBackups, for example. Still thinking through it - interested in your thoughts.

Reply:
Hmm, maybe. This may have implications for how they're currently backed up, but I'm not sure. I'll have to think on that some more. 

As it stands, the JSON file with the VolumeSnapshots listed is just for getting backup details, but it may make sense to sync them in and just query the cluster. The controller would need to sync in the VolumeSnapshotClass, VolumeSnapshotContent, and VolumeSnapshot at sync time.

Also note that VolumeSnapshot objects live inside the same namespace as the PVC they captured from, so we'd also have to recreate the NS(es).

Reply:
Good point, yeah I don't think we want to create the NSes as part of it. I suppose we could create the SnapshotClass + VSCs only, but even that might be weird.

Will ponder some more.  The inconsistencies between current snapshots, restic, and CSI snapshots bother me, but they're also hard to reconcile given the different models that all three of these have.

Reply:
Yeah, thinking about this more, I don't think we want to try to do anything with the sync controller.